### PR TITLE
Refactor matrix_action

### PR DESCRIPTION
**Pull request by nilsnolde**
as a preparation to https://github.com/valhalla/valhalla/issues/4474

similar to routing algos, I created a base class for matrix algos and unified the main public `SourceToTarget` function. IMO the logic is a bit nicer now, but also it'll be much much easier to handle second passes this way.

I should add this is really a pure refactor, no functional change is intended.

### CODE REVIEW

File: src/thor/costmatrix.cc [hunk 1]
```diff
@@ -49,9 +49,9 @@ class CostMatrix::ReachedMap : public robin_hood::unordered_map<uint64_t, std::v
 
 // Constructor with cost threshold.
 CostMatrix::CostMatrix(const boost::property_tree::ptree& config)
-    : max_reserved_labels_count_(config.get<uint32_t>("max_reserved_labels_count_bidir_dijkstras",
+    : MatrixAlgorithm(config),
+      max_reserved_labels_count_(config.get<uint32_t>("max_reserved_labels_count_bidir_dijkstras",
                                                       kInitialEdgeLabelCountBidirDijkstra)),
-      clear_reserved_memory_(config.get<bool>("clear_reserved_memory", false)),
```
Comment (nilsnolde) [inline src/thor/costmatrix.cc:54]: this is now in the base class

File: src/thor/costmatrix.cc [hunk 2]
```diff
@@ -132,19 +132,17 @@ void CostMatrix::SourceToTarget(Api& request,
                                 baldr::GraphReader& graphreader,
                                 const sif::mode_costing_t& mode_costing,
                                 const sif::travel_mode_t mode,
-                                const float max_matrix_distance,
-                                const bool has_time,
-                                const bool invariant,
-                                const ShapeFormat& shape_format) {
```
Comment (nilsnolde) [inline src/thor/costmatrix.cc:138]: I pulled all these out of the signature and set them inside the function instead. was pointless to have them in the signature, it'll all in the `request`, except for `has_time` which is being set via the base class's `set_has_time()`

File: src/thor/matrix_action.cc [hunk 3]
```diff
@@ -21,43 +21,9 @@ constexpr uint32_t kCostMatrixThreshold = 5;
...
-  };
-  auto timedistancematrix = [&]() {
-    if (options.shape_format() != no_shape)
-      add_warning(request, 207);
-    return time_distance_matrix_.SourceToTarget(request, *reader, mode_costing, mode,
-                                                max_matrix_distance.find(costing)->second,
-                                                options.matrix_locations(),
-                                                options.date_time_type() == Options::invariant);
-  };
+MatrixAlgorithm* thor_worker_t::get_matrix_algorithm(Api& request, const bool has_time) {
```
Comment (nilsnolde) [inline src/thor/matrix_action.cc:24]: the diff in here is impossible to read. will show tmrw in code editor what's going on there. the basic gist is:
- add a `get_matrix_algorithm` function to thor_worker_t (like for routing), to get the properties of that class
- remove the lambdas for the matrix algos and decide here which one to use

File: src/thor/matrix_action.cc [hunk 4]
```diff
@@ -81,32 +47,60 @@ std::string thor_worker_t::matrix(Api& request) {
...
+      check_matrix_time(request, options.prioritize_bidirectional() ? Matrix::CostMatrix
+                                                                    : Matrix::TimeDistanceMatrix);
+
+  // allow all algos to be cancelled
+  for (auto* alg : std::vector<MatrixAlgorithm*>{
+           &costmatrix_,
+           &time_distance_matrix_,
+           &time_distance_bss_matrix_,
+       }) {
+    alg->set_interrupt(interrupt);
```
Comment (nilsnolde) [inline src/thor/matrix_action.cc:94]: I added the `interrupt` function now for all matrix algos finally

File: src/thor/timedistancematrix.cc [hunk 5]
```diff
@@ -8,33 +8,14 @@
 using namespace valhalla::baldr;
 using namespace valhalla::sif;
 
-namespace {
-static bool IsTrivial(const uint64_t& edgeid,
-                      const valhalla::Location& origin,
```
Comment (nilsnolde) [inline src/thor/timedistancematrix.cc:13]: this was duplicated with the exact same function in `PathAlgorithm`

File: src/tyr/matrix_serializer.cc [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 #include "baldr/json.h"
 #include "proto_conversions.h"
-#include "thor/matrix_common.h"
```
Comment (nilsnolde) [inline src/tyr/matrix_serializer.cc:5]: I also remove this header and pulled out what was there into the new header

File: test/matrix.cc [hunk 7]
```diff
@@ -220,7 +220,7 @@ TEST(Matrix, test_matrix) {
   CostMatrix cost_matrix;
   cost_matrix.SourceToTarget(request, reader, mode_costing, sif::TravelMode::kDrive, 400000.0);
   auto matrix = request.matrix();
-  for (uint32_t i = 0; i < matrix.times().size(); ++i) {
+  for (int i = 0; i < matrix.times().size(); ++i) {
```
Comment (nilsnolde) [inline test/matrix.cc:223]: made this file warning-free

File: valhalla/thor/pathalgorithm.h [hunk 8]
```diff
@@ -155,32 +155,32 @@ class PathAlgorithm {
...
+/**
+ * Check for path completion along the same edge. Edge ID in question
+ * is along both an origin and destination and origin shows up at the
+ * beginning of the edge while the destination shows up at the end of
+ * the edge.
+ * @param  edgeid       Edge id.
+ * @param  origin       Origin path location information.
+ * @param  destination  Destination path location information.
+ */
+inline bool IsTrivial(const baldr::GraphId& edgeid,
```
Comment (nilsnolde) [inline valhalla/thor/pathalgorithm.h:169]: this is changed from a member function to an inline function

**Comment by nilsnolde** [comment]
Also, in a further step, we could think about somehow merging `PathAlgorithm` with `MatrixAlgorithm` and also add a bit more common functionality around expansion details if possible, so we don't have to maintain so many places doing the same thing. That needs a lot more thought though, but is still kinda my end goal.

File: valhalla/thor/matrixalgorithm.h [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+  // for tracking the expansion of the algorithm visually
+  expansion_callback_t expansion_callback_;
+
+  uint32_t max_reserved_labels_count_;
+
+  // if `true` clean reserved memory for edge labels
+  bool clear_reserved_memory_;
+
+  // resizes all PBF sequences except for date_times
```
Comment (chrstnbwnkl) [inline valhalla/thor/matrixalgorithm.h:116]: stale comment

**Review by kevinkreiser — APPROVED** [review]

**Comment by nilsnolde** [comment]
fixed the build @kevinkreiser , if you have a sec to re-approve

**Review by kevinkreiser — APPROVED** [review]

**Pull request merged**